istioctl/cmd/auth.go

```diff
@@ -234,16 +237,73 @@ func getConfigDumpFromPod(podName, podNamespace string) (*configdump.Wrapper, er
 }
 
 func newConverter(v1PolicyFiles, serviceFiles []string, istioNamespace, istioMeshConfigMapName string) (*converter.Converter, error) {
-	if len(v1PolicyFiles) == 0 {
-		return nil, fmt.Errorf("no input file provided")
+	v1alpha1Policies, err := getV1alpha1Policies(v1PolicyFiles)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get v1alpha1 policies: %v", err)
 	}
 
-	var k8sClient *kubernetes2.Clientset
 	k8sClient, err := kube.CreateClientset("", "")
 	if err != nil {
 		return nil, fmt.Errorf("failed to connect to Kubernetes: %v", err)
 	}
-	return converter.New(k8sClient, v1PolicyFiles, serviceFiles, meshConfig, istioNamespace, istioMeshConfigMapName)
+	return converter.New(k8sClient, v1alpha1Policies, serviceFiles, meshConfig, istioNamespace, istioMeshConfigMapName)
+}
+
+func getV1alpha1Policies(v1PolicyFiles []string) (*model.AuthorizationPolicies, error) {
+	var authzPolicies *model.AuthorizationPolicies
+	if len(v1PolicyFiles) != 0 {
+		var configs []model.Config
+		for _, fileName := range v1PolicyFiles {
+			rbacFileBuf, err := ioutil.ReadFile(fileName)
+			if err != nil {
+				return nil, err
+			}
+			configFromFile, _, err := crd.ParseInputs(string(rbacFileBuf))
+			if err != nil {
+				return nil, err
+			}
+			configs = append(configs, configFromFile...)
+		}
+
+		store := model.MakeIstioStore(memory.Make(schemas.Istio))
+		for _, config := range configs {
+			if _, err := store.Create(config); err != nil {
+				return nil, err
+			}
+		}
+		env := &model.Environment{
+			IstioConfigStore: store,
+		}
+		var err error
+		authzPolicies, err = model.GetAuthorizationPolicies(env)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		kubeClient, err := clientExecFactory(kubeconfig, configContext)
+		if err != nil {
+			return nil, err
+		}
+		results, err := kubeClient.AllPilotsDiscoveryDo(istioNamespace, "GET", "/debug/authorizationz", nil)
```

Comment:
Just curious, why not reading directly from api server to get authz policies, rather than exec-ing to the pilot-discovery container?

Reply:
1. Much harder to get CRDs from k8s API server; 2. reading from pilot gives the up-to-date `AuthorizationPolicy` and some other useful information, e.g. root namespace.

Reply:
I'm curious why is it harder to get CRDs from k8s API server? Is it due to permissions?

Reply:
I think that requires some quire extra setting for the kubeclient which is not used in istioctl.